packages/nfts/contracts/party-ticket/TaikoPartyTicket.sol

```diff
@@ -0,0 +1,229 @@
+    function setWinners(uint256[] calldata _winners)
+        external
+        whenNotPaused
+        onlyRole(DEFAULT_ADMIN_ROLE)
+    {
+        for (uint256 i = 0; i < _winners.length; i++) {
+            winners[_winners[i]] = true;
+        }
+        pause();
+    }
+
+    /// @notice Set the base URI
+    /// @param _baseURI The new base URI
+    function setBaseURI(string memory _baseURI) external onlyRole(DEFAULT_ADMIN_ROLE) {
+        baseURI = _baseURI;
+    }
+
+    /// @notice Set the winner base URI
+    /// @param _winnerBaseURI The new winner base URI
+    function setWinnerURI(string memory _winnerBaseURI) external onlyRole(DEFAULT_ADMIN_ROLE) {
+        winnerBaseURI = _winnerBaseURI;
+    }
+
+    /// @notice Mint a raffle ticket
+    /// @dev Requires a fee to mint
+    /// @dev Requires the contract to not be paused
+    function mint() external payable whenNotPaused {
+        if (blacklist.isBlacklisted(_msgSender())) revert ADDRESS_BLACKLISTED();
+        if (msg.value < mintFee) revert INSUFFICIENT_MINT_FEE();
+        uint256 tokenId = _nextTokenId++;
+        _safeMint(msg.sender, tokenId);
+    }
+
+    /// @notice Mint multiple raffle tickets
+    /// @param amount The number of tickets to mint
+    /// @dev Requires a fee to mint
+    /// @dev Requires the contract to not be paused
+    function mint(uint256 amount) external payable whenNotPaused {
+        if (blacklist.isBlacklisted(_msgSender())) revert ADDRESS_BLACKLISTED();
+        if (msg.value < mintFee * amount) revert INSUFFICIENT_MINT_FEE();
+        for (uint256 i = 0; i < amount; i++) {
+            uint256 tokenId = _nextTokenId++;
+            _safeMint(msg.sender, tokenId);
+        }
+    }
+
+    /// @notice Mint a raffle ticket
+    /// @param to The address to mint to
+    /// @dev Requires the contract to not be paused
+    /// @dev Can only be called by the admin
+    function mint(address to) public whenNotPaused onlyRole(DEFAULT_ADMIN_ROLE) {
+        if (blacklist.isBlacklisted(to)) revert ADDRESS_BLACKLISTED();
+        uint256 tokenId = _nextTokenId++;
+        _safeMint(to, tokenId);
+    }
+
+    /// @notice Mint a winner ticket
+    /// @param to The address to mint to
+    /// @dev Requires calling as an admin
+    function mintWinner(address to) public onlyRole(DEFAULT_ADMIN_ROLE) {
+        if (blacklist.isBlacklisted(to)) revert ADDRESS_BLACKLISTED();
+        uint256 tokenId = _nextTokenId++;
+        winners[tokenId] = true;
+        _safeMint(to, tokenId);
+    }
+
+    /// @notice Revoke a winner and replace with a new winner
+    /// @param revokeId The ID of the winner to revoke
+    /// @param newWinnerId The ID of the new winner
+    function revokeAndReplaceWinner(
+        uint256 revokeId,
+        uint256 newWinnerId
+    )
+        external
+        whenPaused
+        onlyRole(DEFAULT_ADMIN_ROLE)
+    {
+        if (!winners[newWinnerId]) revert CANNOT_REVOKE_NON_WINNER();
```

Comment:
how about create a `revokeWinners`, which can work with `setWinners` to replace `revokeAndReplaceWinner`?

Reply:
Adding the `revokeWinner` method. I'd leave in the `revokeAndReplaceWinner` to ease hot-fixing the winners list.